website/addons/s3/views.py

```diff
@@ -0,0 +1,189 @@
+from modularodm.storage.base import KeyExistsException
+
+from framework.exceptions import HTTPError
+from framework.auth.decorators import must_be_logged_in
+
+from website.addons.base import generic_views
+from website.addons.s3 import utils
+from website.addons.s3.serializer import S3Serializer
+from website.oauth.models import ExternalAccount
+from website.project.decorators import (
+    must_have_addon, must_have_permission,
+    must_not_be_registration, must_be_addon_authorizer,
+)
+
+SHORT_NAME = 's3'
+FULL_NAME = 'Amazon S3'
+
+s3_account_list = generic_views.account_list(
+    SHORT_NAME,
+    S3Serializer
+)
+
+s3_import_auth = generic_views.import_auth(
+    SHORT_NAME,
+    S3Serializer
+)
+
+s3_deauthorize_node = generic_views.deauthorize_node(
+    SHORT_NAME
+)
+
+s3_get_config = generic_views.get_config(
+    SHORT_NAME,
+    S3Serializer
+)
+
+def _get_buckets(node_addon, folder_id=None):
+    """Used by generic_view `folder_list` to fetch a list of buckets.
+    `folder_id` required by generic, but not actually used"""
+    return {
+        'buckets': utils.get_bucket_names(node_addon)
+    }
+
+s3_folder_list = generic_views.folder_list(
+    SHORT_NAME,
+    FULL_NAME,
+    _get_buckets
+)
+
+s3_root_folder = generic_views.root_folder(
+    SHORT_NAME
+)
+
+@must_be_logged_in
+def s3_add_user_account(auth, **kwargs):
+    """Verifies new external account credentials and adds to user's list"""
+    try:
+        access_key = request.json['access_key']
+        secret_key = request.json['secret_key']
+    except KeyError:
+        raise HTTPError(httplib.BAD_REQUEST)
+
+    if not (access_key and secret_key):
+        return {
+            'message': ('All the fields above are required.')
+        }, httplib.BAD_REQUEST
+
+    user_info = utils.get_user_info(access_key, secret_key)
+    if not user_info:
+        return {
+            'message': ('Unable to access account.\n'
+                'Check to make sure that the above credentials are valid, '
+                'and that they have permission to list buckets.')
+        }, httplib.BAD_REQUEST
+
+    if not utils.can_list(access_key, secret_key):
+        return {
+            'message': ('Unable to list buckets.\n'
+                'Listing buckets is required permission that can be changed via IAM')
+        }, httplib.BAD_REQUEST
+
+    account = None
+    try:
+        account = ExternalAccount(
+            provider=SHORT_NAME,
+            provider_name=FULL_NAME,
+            oauth_key=access_key,
+            oauth_secret=secret_key,
+            provider_id=user_info.id,
+            display_name=user_info.display_name,
+        )
+        account.save()
+    except KeyExistsException:
+        # ... or get the old one
+        account = ExternalAccount.find_one(
+            Q('oauth_key', 'eq', access_key) &
+            Q('oauth_secret', 'eq', secret_key)
+        )
+    assert account is not None
+
+    if account not in auth.user.external_accounts:
```

Comment:
If `external_accounts` is a foreign field this might not work all the time as @cwisecarver and myself found out.